GS: Add gamma control to ShadeBoost

Gonzalosilvalde opened the pull request:
### Description of Changes
<!-- Brief description or overview on what was changed in the PR -->
- Add a Gamma parameter to ShadeBoost shaders (OpenGL & Vulkan).
- Expose a Gamma slider in the Qt Graphics Settings and the fullscreen in-game menu.
### Rationale behind Changes
<!-- Why were these changes made?  What problem does it solve / area does it improve? -->
Gamma control lets users adjust mid-tones independently of brightness/contrast/saturation. Fulfills feature request #12515.
### Suggested Testing Steps
<!-- If applicable, including examples you've already tested with / recommendations for how to test further is very helpful! -->
- Open PCSX2.
- Emulate a game.
- Open graphics settings.
- Go to Postprocessing.
- Adjust Gamma as you see fit.

RedDevilus commented:
Don't forget to format your files correctly, there are spaces where there have to be tabs, like for example:

![image](https://github.com/user-attachments/assets/e143b9b9-a177-4d13-acc7-5e8b6e35d6ba)


(Using refined github extension in my browser to easily detect)

Gonzalosilvalde commented:
I had a doubt about that, I saw those weird spacing in the git diff but in my text editor don't appear.

JordanTheToaster commented:
> I had a doubt about that, I saw those weird spacing in the git diff but in my text editor don't appear.

You are using spaces not tabs.

Gonzalosilvalde commented:
I still have errors...

Gonzalosilvalde commented:
I think it's ok now, sorry for the problems.

JordanTheToaster commented:
Do you plan to add these to Direct3D 11 and 12?

Gonzalosilvalde commented:
I hadn’t planned on it, but I have no problem doing it if you’d like. At a glance, it should be pretty much the same.

JordanTheToaster commented:
It would be good to have it on all backends if possible.

Gonzalosilvalde commented:
I believe this change is sufficient for Direct3D 11 and 12, but I have no way to verify that it works. I would appreciate it if someone could test it so I can make any necessary adjustments.

KrossX commented:
Tested on the Windows side, the following changes makes the dx11 shader changes work.
```diff
 	float3 satColor = lerp(intensity, brtColor, sat);
 	float3 conColor = lerp(AvgLumin, satColor, con);
 	
-	vec3 csb = conColor;
-	csb = pow(csb, vec3(1.0 / gam));
+	float gam_inv = 1.0/gam;
+	float3 csb = conColor;
+	csb = pow(csb, float3(gam_inv, gam_inv, gam_inv));
 	color.rgb = csb;
 	return color;
```

Also, the new GUI control does not follow the enable/disable state from the shadeboost checkbox. Other than that, it seems to work. 👍

lightningterror commented on bin/resources/shaders/dx11/shadeboost.fx:
```diff
@@ -34,8 +35,10 @@ float4 ContrastSaturationBrightness(float4 color) // Ported to HLSL
-	color.rgb = conColor;	
+	
+	vec3 csb = conColor;
+	csb = pow(csb, vec3(1.0 / gam));
```
float3, same for the above.

lightningterror reviewed (commented):
Metal remaining.

AmyRoxwell commented:
Minor bug (idk if it's exclusive to linux) but probably would be a good idea to position these buttons better on the blank space we have free. Currently it looks like this on 1080p screen
![Screenshot_20250512_164635](https://github.com/user-attachments/assets/56e85c59-cbbb-4675-9e4a-4e414d4f3bef)

KrossX commented:
That might be good for another PR, to make that tab more organized. There's probably enough room for nice sliders too while at it.

Gonzalosilvalde commented:
## In this last commit I have changed:
- Now you can enable/disable the state from the `shade boost` box.
- In dx11/shadeboost.fx use a float3 and not a vec3 (needs test).
- Added gamma in metal (needs test).
- All 'shade boost' options are in alphabetical order. 

## What I didn't do:
- I don't know what @Mrlinkwii is referring to when he says “shadercache version needs bumped”, if someone explains it to me I do it no problem.
- I didn't position the buttons better in the white space as @AmyRoxwell suggested. As @KrossX said, I think it should go in another PR.

As is obvious, if any problems are encountered whether in dx11/12 or metal I'll be glad to fix them.

kamfretoz commented:
>     * I don't know what @Mrlinkwii is referring to when he says “shadercache version needs bumped”, if someone explains it to me I do it no problem.

https://github.com/PCSX2/pcsx2/blob/master/pcsx2/ShaderCacheVersion.h#L6

You'll want to increment the shader cache version by one.

Gonzalosilvalde commented:
sorry for taking so long to make the change, I didn't have time to do it or I got distracted.

I have one thing to comment after the rebase.

In the functionalities that are similar in pcsx2/ImGui/FullscreenUI.cpp, now FSUI_ICONSTR is used, I kept FSUI_CSTR in the gamma control because I didn't know which ICON to put it, since I saw that in the others I had ICON_FA_GEM, ICON_FA_SUN... and the function needs one of these ICONs

TheTechnician27 commented:
<img width="1136" height="209" alt="image" src="https://github.com/user-attachments/assets/bd37b5ff-0c55-465d-8d07-6fb98e0cd249" />

<img width="605" height="121" alt="image" src="https://github.com/user-attachments/assets/f679105b-cb88-455b-85a0-6585c41bad8f" />

A default integer for the gamma value is never set, and consequently it defaults to '1' in master.

Also, why is 'BindWidgetToIntSetting' being used for a bool in four of these (where one of them comes from this PR)?
